File: frontend/lib/backend/agents/affordability-agent.ts
import type { AgentReport, Applicant, Contributor } from "../schemas";

export function runAffordabilityAgent(applicant: Applicant): AgentReport {
  const signals = applicant.document_signals ?? {};
  const numericHints = signals.numeric_hints ?? {};

  const effectiveIncome = Math.min(
    applicant.monthly_income,
    numericHints.monthly_income ?? applicant.monthly_income
  );
  const effectiveExpenses = Math.max(
    applicant.monthly_expenses,
    numericHints.monthly_expenses ?? applicant.monthly_expenses
  );
  const effectiveDebt = Math.max(
    applicant.existing_debt,
    numericHints.existing_debt ?? applicant.existing_debt
  );

  const freeCashFlow = effectiveIncome - effectiveExpenses;
  const dti = effectiveDebt / Math.max(effectiveIncome, 1);
  const installmentProxy = applicant.requested_amount / 36;
  const burden = installmentProxy / Math.max(freeCashFlow, 1);

  let risk =
    0.15 +
    Math.min(dti, 1.5) * 0.35 +
    Math.min(burden, 2.0) * 0.25 +
    Math.min(applicant.overdrafts_90d, 5) * 0.04;

  if (applicant.income_verified) risk -= 0.08;
  if (freeCashFlow > installmentProxy * 2) risk -= 0.1;

  const score = Math.max(0, Math.min(1, 1 - risk));

  const contributors: Contributor[] = [
    {
      feature: "debt_to_income_ratio",
      impact: Math.round(Math.min(dti, 1.5) * 0.35 * 100) / 100,
      direction: "increases risk",
      explanation: `Debt-to-income ratio is ${dti.toFixed(2)}.`,
    },
    {
      feature: "free_cash_flow",
      impact:
        freeCashFlow > installmentProxy * 2
          ? -0.1
          : 0.12,
      direction:
        freeCashFlow > installmentProxy * 2
          ? "decreases risk"
          : "increases risk",
      explanation: `Estimated free cash flow is $${freeCashFlow.toFixed(0)}.`,
    },
    {
      feature: "overdrafts_90d",
      impact: Math.round(Math.min(applicant.overdrafts_90d, 5) * 0.04 * 100) / 100,
      direction: "increases risk",
      explanation: `Overdraft count in last 90 days: ${applicant.overdrafts_90d}.`,
    },
  ];

  const reasons: string[] = [];
  if (dti > 0.45) reasons.push("HIGH_DTI");
  if (burden > 0.6) reasons.push("LOW_AFFORDABILITY_BUFFER");
  if (applicant.overdrafts_90d >= 2) reasons.push("RECENT_OVERDRAFTS");
  if (Object.keys(numericHints).length) reasons.push("DOC_HINTS_APPLIED");
  if (!reasons.length) reasons.push("AFFORDABILITY_BUFFER_OK");

  const recommendation =
    score >= 0.72 ? "PASS" : score >= 0.45 ? "REFER" : "REJECT";

  return {
    agent_name: "Affordability Agent",
    score: Math.round(score * 100) / 100,
    model_version: "affordability-monotonic-xgb-v1",
    top_contributors: contributors,
    evidence_refs: ["bank_statement", "income_proof"],
    reason_codes: reasons,
    confidence_status: "demo_calibrated",
    recommendation,
    summary:
      "Monotonic affordability model: risk rises with DTI, burden, overdrafts and falls with verified income/free cash flow.",
  };
}

File: backend/app/agents/affordability_agent.py
from app.models.schemas import SpendProfile, AgentReport, Contributor


def run(profile: SpendProfile) -> AgentReport:
    signals = profile.document_signals or {}
    numeric_hints = signals.get("numeric_hints", {})

    effective_revenue = min(profile.monthly_revenue, float(numeric_hints.get("monthly_revenue", profile.monthly_revenue)))
    effective_spend = max(profile.monthly_spend, float(numeric_hints.get("monthly_spend", profile.monthly_spend)))
    effective_budget = min(profile.planned_budget, float(numeric_hints.get("planned_budget", profile.planned_budget)))

    efficiency_margin = (effective_revenue - effective_spend) / max(effective_revenue, 1)
    budget_overshoot = max(0.0, (effective_spend - effective_budget) / max(effective_budget, 1))
    reserve_buffer = profile.cash_reserve / max(effective_spend, 1)

    risk = 0.20 + min(budget_overshoot, 1.5) * 0.35 + (0.25 if efficiency_margin < 0 else 0.08)
    risk += min(profile.late_payments_90d, 5) * 0.05
    if profile.books_verified:
        risk -= 0.07
    if reserve_buffer > 1.2:
        risk -= 0.12
    score = max(0.0, min(1.0, 1 - risk))

    contributors = [
        Contributor(feature="budget_overshoot", impact=round(min(budget_overshoot, 1.5) * 0.35,2), direction="increases risk", explanation=f"Spend-to-budget overshoot ratio is {budget_overshoot:.2f}."),
        Contributor(feature="efficiency_margin", impact=round(-0.10 if efficiency_margin > 0.2 else 0.16,2), direction="decreases risk" if efficiency_margin > 0.2 else "increases risk", explanation=f"Operating spend margin is {efficiency_margin:.2%}."),
        Contributor(feature="late_payments_90d", impact=round(min(profile.late_payments_90d,5)*0.05,2), direction="increases risk", explanation=f"Late vendor payments in last 90 days: {profile.late_payments_90d}."),
    ]
    reasons=[]
    if budget_overshoot > 0.15: reasons.append("BUDGET_OVERSHOOT")
    if efficiency_margin < 0.05: reasons.append("LOW_SPEND_EFFICIENCY")
    if profile.late_payments_90d >= 2: reasons.append("RECENT_LATE_PAYMENTS")
    if numeric_hints:
        reasons.append("DOC_HINTS_APPLIED")
    if not reasons: reasons=["SPEND_EFFICIENCY_STABLE"]
    recommendation = "HEALTHY" if score >= .72 else "WATCHLIST" if score >= .45 else "ACTION_REQUIRED"
    return AgentReport(agent_name="Spend Efficiency Agent", score=round(score,2), model_version="spend-efficiency-monotonic-v2", top_contributors=contributors, evidence_refs=["general_ledger", "budget_plan"], reason_codes=reasons, confidence_status="demo_calibrated", recommendation=recommendation, summary="Evaluates budget discipline, spend efficiency margin, and payment hygiene to score operational spend quality.")
